migration-engine/connectors/sql-migration-connector/src/flavour.rs

```diff
@@ -15,45 +16,101 @@ pub(crate) use sqlite::SqliteFlavour;
+    sql_schema_differ::SqlSchemaDifferFlavour,
 };
 use datamodel::{common::preview_features::PreviewFeature, ValidatedSchema};
-use enumflags2::BitFlags;
-use migration_connector::{migrations_directory::MigrationDirectory, ConnectorError, ConnectorResult};
-use quaint::prelude::{ConnectionInfo, Table};
+use migration_connector::{
+    migrations_directory::MigrationDirectory, BoxFuture, ConnectorError, ConnectorParams, ConnectorResult,
+};
+use quaint::prelude::Table;
 use sql_schema_describer::SqlSchema;
 use std::fmt::Debug;
 use user_facing_errors::migration_engine::ApplyMigrationError;
 
-pub(crate) fn from_connection_info(
-    connection_info: &ConnectionInfo,
-    preview_features: BitFlags<PreviewFeature>,
-) -> Box<dyn SqlFlavour + Send + Sync + 'static> {
-    match connection_info {
-        ConnectionInfo::Mysql(url) => Box::new(MysqlFlavour::new(url.clone(), preview_features)),
-        ConnectionInfo::Postgres(url) => Box::new(PostgresFlavour::new(url.clone(), preview_features)),
-        ConnectionInfo::Sqlite { file_path, db_name } => Box::new(SqliteFlavour {
-            file_path: file_path.clone(),
-            attached_name: db_name.clone(),
-            preview_features,
-        }),
-        ConnectionInfo::Mssql(url) => Box::new(MssqlFlavour::new(url.clone(), preview_features)),
-        ConnectionInfo::InMemorySqlite { .. } => unreachable!("SqlFlavour for in-memory SQLite"),
+/// P is the params, C is a connection.
+pub(crate) enum State<P, C> {
+    Initial,
+    WithParams(P),
+    Connected(P, C),
+}
+
+impl<P, C> Debug for State<P, C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            State::Initial => f.write_str("State::Initial"),
+            State::WithParams(_) => f.write_str("State::Params(<CONFIDENTIAL>)"),
+            State::Connected(_, _) => f.write_str("State::Connected(<CONFIDENTIAL>)"),
+        }
+    }
+}
+
+impl<P, C> State<P, C>
+where
+    P: 'static,
+    C: 'static,
+{
+    fn params(&self) -> Option<&P> {
+        match self {
+            State::Initial => None,
+            State::WithParams(p) | State::Connected(p, _) => Some(p),
+        }
+    }
+
+    /// Unwrap the state's params. We do not return an error because we want to trigger a panic if
+    /// that happens, because it means an internal logic error.
+    ///
+    /// This is useful when you want the params, but you do not care if a connection has been
+    /// started or not.
+    #[track_caller]
+    fn get_unwrapped_params(&self) -> &P {
+        match self {
+            State::Initial => panic!("Internal logic error: get_unwrapped_params() on State::Initial"),
+            State::WithParams(p) => p,
+            State::Connected(p, _) => p,
+        }
+    }
+
+    #[track_caller]
+    fn set_params(&mut self, params: P) {
+        match self {
+            State::WithParams(_) | State::Connected(_, _) => panic!("state error"),
```

Comment:
Instead of state errors, could we enforce the usage of the state machine with the generic types? Just implement `set_params` if the `Connection` and `Params` are something like `()`. When setting, return a type that has actual structs in the generic types and this one doesn't have `set_params` anymore.